spec/consensus/reactor/reactor.md

```diff
@@ -0,0 +1,355 @@
+What is required is for either the message to be delivered or that, eventually, some newer message, with information that supersedes that in the first message, to be timely delivered.
+In Tendermint BFT, this property is seen, for example, when nodes ignore proposal messages from prior rounds.
+
+> **Supersession**: We say that a message $lhs$ supersedes a message $rhs$ if after receiving $lhs$ a process would make at least as much progress as it would by receiving $rhs$ and we note is as follows: $\text{lhs} \overset{s}{>} \text{rhs}$.   
+> **TODO**: Better definition.
+
+It seems reasonable, therefore, to formalize the requirements of Tendermint in terms of a weaker, *best-effort* communication guarantees and to combine them with GST outside of the communication layer, to ensure eventual termination.
+
+> **Note**   
+> It is also assumed that, after GST, timeouts do not expire precociously.
+
+
+
+
+
+# Part 2: Specifications
+## The Consensus State Layer - State
+
+### Provides to Applications and other Reactors
+
+Communication of the Consensus Reactor with layers above (applications) and on to the side (other reactors) is covered by the [ABCI](../../abci/) specification.
+
+### Provides to the Communication Layer
+
+A message supersession operator:
+
+`SSS(lhs,rhs)` (`lhs` supersedes `rhs`) is true if and only if $\text{lhs} \overset{s}{>} \text{rhs}$
+
+### Requires from the Communication Layer
+
+Messages received by a node either change the node's State or are discarded. 
+If they change the State, even if by just getting added to a set of received messages, then the original message need not be forwarded. 
+Instead, a new message, with the same contents, is generated from the State of the node. Why does this matter? Because it let's us think in terms of messages being sent to neighbors only, not forwarded on the overlay.
+
+> **[REQ-STATE-GOSSIP-NEIGHBOR_CAST]**   
+> Best Effort Neighbor-Cast: continuously resend any non-superseded messages to neighbors (connected nodes).
+
+It should be provable that    
+[REQ-STATE-GOSSIP-NEIGHBOR_CAST] +  $\Diamond\Delta$-Timely Communication ~> $\Diamond$ Termination.
+
+
+
+## Communication, AKA Gossip
+
+
+```tla+
+SuperS(lhs,rhs): the supersession operator  
+Ne[p]: neighbor set of p
+Msgs[p]: the messages for which GossipCast has been invoked at p
+DMsgs[p]: set of messages GossipDelivered by p
+```
+
+### Provides to State (Gossip-I)
+
+[GOSSIP-CAST-NEIGHBORS]
+
+To gossip-cast, add message to the outbound messages set.
+
+```tla+
+GossipCast(p,m): Msgs[p]' = Msgs[p] \cup {m} 
+```
+
+All outbound messages will be delivered to neighbors at the time the message was sent, or the neighbor disconnects, or a superseding messages makes sending the message useless.
+
+```tla+
+GossipCastDelivery(p,q,m) == 
+    q \in Ne[p] /\ m \in Msgs[p] ~> \/ m \in DMsgs[q]
+                                    \/ m \notin Msgs[p] /\ \E mn \in Msgs[p]: SuperS(mn,m)
```

Comment:
```suggestion
                                    \/ \E mn \in Msgs[p]: SuperS(mn,m)
```
It seems to me that logically the "\notin" part is not needed

Reply:
True. We don't need to lose the message, just not deliver it.